File: polymarket/data_pipeline/weatherapi_fetcher.py
"""
WeatherAPI.com data fetcher module
Retrieves historical weather data from WeatherAPI.com
Requires API key, global coverage
"""

import logging
import requests
import pandas as pd
from datetime import datetime, timedelta
from typing import List, Optional

from polymarket.config import WEATHERAPI_KEY

logger = logging.getLogger(__name__)

LOCATIONS = {
    "NYC": {"query": "New York", "name": "New York, NY"},
    "LA": {"query": "Los Angeles", "name": "Los Angeles, CA"},
    "London": {"query": "London", "name": "London, UK"},
    "Chicago": {"query": "Chicago", "name": "Chicago, IL"},
    "Dallas": {"query": "Dallas", "name": "Dallas, TX"},
    "Denver": {"query": "Denver", "name": "Denver, CO"},
    "Miami": {"query": "Miami", "name": "Miami, FL"},
    "Boston": {"query": "Boston", "name": "Boston, MA"},
}

BASE_URL = "https://api.weatherapi.com/v1/history.json"


class WeatherAPIFetcher:
    """Fetches historical weather data from WeatherAPI.com"""

    def __init__(self, api_key: str = WEATHERAPI_KEY):
        self.api_key = api_key
        self.session = requests.Session()

    def is_available(self) -> bool:
        """Check if API key is configured"""
        return bool(self.api_key)

    def fetch_daily_observations(
        self,
        query: str,
        start_date: str,
        end_date: str,
        location_id: str = "WAPI",
    ) -> pd.DataFrame:
        """
        Fetch daily weather observations for a location

        Args:
            query: Location query string
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)
            location_id: Identifier for the location

        Returns:
            DataFrame with standardized weather observations
        """
        if not self.is_available():
            logger.info("WeatherAPI: No API key configured, skipping")
            return pd.DataFrame()

        start = datetime.strptime(start_date, "%Y-%m-%d")
        end = datetime.strptime(end_date, "%Y-%m-%d")
        observations = []

        current = start
        while current <= end:
            date_str = current.strftime("%Y-%m-%d")
            params = {
                "key": self.api_key,
                "q": query,
                "dt": date_str,
            }

            try:
                response = self.session.get(BASE_URL, params=params, timeout=15)
                response.raise_for_status()
                data = response.json()

                forecast = data.get("forecast", {}).get("forecastday", [])
                if forecast:
                    day = forecast[0].get("day", {})
                    observations.append({
                        "date": pd.Timestamp(date_str),
                        "station_id": f"WAPI_{location_id}",
                        "temperature_max": day.get("maxtemp_c"),
                        "temperature_min": day.get("mintemp_c"),
                        "temperature_mean": day.get("avgtemp_c"),
                        "precipitation_total": day.get("totalprecip_mm", 0),
                        "wind_speed_mean": day.get("maxwind_kph", 0) / 3.6,  # km/h to m/s
                    })

            except requests.RequestException as e:
                logger.debug(f"WeatherAPI error for {date_str}: {e}")

            current += timedelta(days=1)

        if not observations:
            return pd.DataFrame()

        df = pd.DataFrame(observations)
        logger.info(f"WeatherAPI: Fetched {len(df)} observations for {location_id}")
        return df

    def fetch_location(
        self,
        location_key: str,
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """Fetch data for a pre-configured location"""
        location = LOCATIONS.get(location_key)
        if not location:
            logger.error(f"Unknown location: {location_key}")
            return pd.DataFrame()

        return self.fetch_daily_observations(
            query=location["query"],
            start_date=start_date,
            end_date=end_date,
            location_id=location_key,
        )

    def fetch_multiple_locations(
        self,
        location_keys: List[str],
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """Fetch data for multiple locations and combine"""
        all_data = []
        for key in location_keys:
            df = self.fetch_location(key, start_date, end_date)
            if not df.empty:
                all_data.append(df)

        if all_data:
            return pd.concat(all_data, ignore_index=True)
        return pd.DataFrame()

File: src/data_pipeline/weatherapi_fetcher.rs
use std::collections::HashMap;

use chrono::{Duration, NaiveDate};
use serde_json::Value;

use crate::config::weatherapi_key;
use crate::types::WeatherRecord;

const BASE_URL: &str = "https://api.weatherapi.com/v1/history.json";

#[derive(Clone)]
pub struct WeatherAPIFetcher {
    api_key: String,
    client: reqwest::blocking::Client,
    locations: HashMap<String, String>,
}

impl WeatherAPIFetcher {
    pub fn new(api_key: Option<String>) -> Self {
        Self {
            api_key: api_key.unwrap_or_else(weatherapi_key),
            client: crate::data_pipeline::build_client(15),
            locations: default_locations(),
        }
    }

    pub fn is_available(&self) -> bool {
        !self.api_key.is_empty()
    }

    pub fn fetch_daily_observations(
        &self,
        query: &str,
        start_date: NaiveDate,
        end_date: NaiveDate,
        location_id: &str,
    ) -> Vec<WeatherRecord> {
        if !self.is_available() {
            return Vec::new();
        }

        let mut out = Vec::new();
        let mut current = start_date;

        while current <= end_date {
            let params = [
                ("key", self.api_key.clone()),
                ("q", query.to_string()),
                ("dt", current.to_string()),
            ];

            if let Ok(resp) = self.client.get(BASE_URL).query(&params).send() {
                if let Ok(json) = resp.json::<Value>() {
                    if let Some(day) = json
                        .get("forecast")
                        .and_then(|v| v.get("forecastday"))
                        .and_then(|v| v.as_array())
                        .and_then(|arr| arr.first())
                        .and_then(|v| v.get("day"))
                    {
                        let mut row = WeatherRecord::new(current, format!("WAPI_{location_id}"));
                        row.temperature_max = day.get("maxtemp_c").and_then(|v| v.as_f64());
                        row.temperature_min = day.get("mintemp_c").and_then(|v| v.as_f64());
                        row.temperature_mean = day.get("avgtemp_c").and_then(|v| v.as_f64());
                        row.precipitation_total = day
                            .get("totalprecip_mm")
                            .and_then(|v| v.as_f64())
                            .or(Some(0.0));
                        row.wind_speed_mean = day
                            .get("maxwind_kph")
                            .and_then(|v| v.as_f64())
                            .map(|x| x / 3.6);
                        out.push(row);
                    }
                }
            }

            current += Duration::days(1);
        }

        out
    }

    pub fn fetch_location(&self, location_key: &str, start_date: NaiveDate, end_date: NaiveDate) -> Vec<WeatherRecord> {
        let Some(query) = self.locations.get(location_key) else {
            return Vec::new();
        };
        self.fetch_daily_observations(query, start_date, end_date, location_key)
    }
}

impl Default for WeatherAPIFetcher {
    fn default() -> Self {
        Self::new(None)
    }
}

fn default_locations() -> HashMap<String, String> {
    [
        ("NYC", "New York"),
        ("LA", "Los Angeles"),
        ("London", "London"),
        ("Chicago", "Chicago"),
        ("Dallas", "Dallas"),
        ("Denver", "Denver"),
        ("Miami", "Miami"),
        ("Boston", "Boston"),
    ]
    .into_iter()
    .map(|(k, v)| (k.to_string(), v.to_string()))
    .collect()
}
